Lecture4/Task4/Voucher.java

```diff
@@ -0,0 +1,65 @@
+import java.util.Locale;
+
+import predefinedExceptions.RegexException;
+
+abstract class Voucher {
+
+    private String name;
+    private double price;
+    private String country;
+    private String startDate;
+    private int countDays;
+    private int discount;
+    private double cost;
+    private String formattedCost;
+
+    Voucher(String[] voucherDetails) throws RegexException {
+        this.name = voucherDetails[1];
+        this.price = Double.parseDouble(voucherDetails[2]);
+        if (voucherDetails[3].matches("[A-Z][A-Z]|[A-Z][A-Z][A-Z]")) {
```

Comment:
1)you now load vouchers from file - that is trusted sources. You don't need to verify the input. Just assume that structure of the file is correct and you need only to grab data.
2)Even for this case, further don't put too much code to constructor.
3) every if-else block can be extracted to a separate method
5) Complex boolean conditions in if, such like this "voucherDetails[4].matches("(0?[1-9]|[12][0-9]|3[01])[.](0?[1-9]|1[012])[.]\\d{4}")" can be extracted to private method like isDate(). The code will be reformatted to if (isDate()) {...} - much simplier

Reply:
fixed in Tasks1-4-fixed package